semtypes/src/main/java/io/ballerina/semtype/Common.java

```diff
@@ -0,0 +1,128 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.semtype;
+
+import io.ballerina.semtype.subtypedata.BddBoolean;
+import io.ballerina.semtype.subtypedata.BddNode;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Code common to implementation of multiple basic types.
+ *
+ * @since 2.0.0
+ */
+public class Common {
+
+    public static boolean typeListIsReadOnly(List<SemType> list) {
+        for (SemType t : list) {
+            if (!Core.isReadOnly(t)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static List<SemType> readOnlyTypeList(List<SemType> mt) {
+        List<SemType> types = new ArrayList<>();
+        for (SemType s : mt) {
+            SemType t;
+            if (Core.isReadOnly(s)) {
+                t = s;
+            } else {
+                t = Core.intersect(s, PredefinedType.READONLY);
+            }
+            types.add(t);
+        }
+        return Collections.unmodifiableList(types);
+    }
+
+    // [from nballerina] A Bdd represents a disjunction of conjunctions of atoms, where each atom is either positive or
+    // negative (negated). Each path from the root to a leaf that is true represents one of the conjunctions
+    // We walk the tree, accumulating the positive and negative conjunctions for a path as we go.
+    // When we get to a leaf that is true, we apply the predicate to the accumulated conjunctions.
+
+    public static boolean bddEvery(TypeCheckContext tc, Bdd b, Conjunction pos, Conjunction neg, Method predicate)
```

Comment:
Shall we use a functional interface like this https://github.com/ballerina-platform/ballerina-lang/pull/32447/files#diff-6f098f688b764c3781866b1ae6ad7f8fb94c4c6bd0b1772bf6587b57001ba41fR44

Reply:
done